keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -199,15 +204,49 @@ open class SqlResourceRepository(
       .execute()
   }
 
+  override fun applicationEventHistory(application: String, limit: Int): List<ApplicationEvent> {
+    require(limit > 0) { "limit must be a positive integer" }
+    return sqlRetry.withRetry(READ) {
+      jooq
+        .select(EVENT.JSON)
+        .from(EVENT)
+        .where(EVENT.SCOPE.eq(Scope.APPLICATION.name))
+        .and(EVENT.UID.eq(application))
+        .orderBy(EVENT.TIMESTAMP.desc())
+        .limit(limit)
+        .fetch()
+        .map { (json) ->
+          objectMapper.readValue<ApplicationEvent>(json)
+        }
+    }
+  }
+
+  override fun applicationEventHistory(application: String, downTo: Instant): List<ApplicationEvent> {
+    return sqlRetry.withRetry(READ) {
+      jooq
+        .select(EVENT.JSON)
+        .from(EVENT)
+        .where(EVENT.SCOPE.eq(Scope.APPLICATION.name))
+        .and(EVENT.UID.eq(application))
+        .and(EVENT.TIMESTAMP.lessOrEqual(LocalDateTime.ofInstant(downTo, ZoneOffset.UTC)))
```

Comment:
Apparently Jackson is converting all timestamps to UTC so JSON is stored that way. Also, for reasons I've been unable to explain, MySQL is _also_ converting timestamps to UTC, even when the timezone it's configured to use is different, so I hard-coded UTC here for that reason.

Reply:
Timestamps are type `java.time.Instant` which is always UTC